### PR TITLE
Implement `LegoCarRaceActor::VTable0x70` and `VTable0x1c`

### PR DESCRIPTION
**Pull request by jonschz**
`VTable0x70` matches to 100 %.

I am somewhat stuck on `VTable0x1c` - the logic seems to be correct, but the stack layout does not quite match and I have run out of ideas; some variables are off by 4 bytes. Suggestions are greatly appreciated.

### CODE REVIEW

File: LEGO1/lego/legoomni/src/race/legoracespecial.cpp [hunk 1]
```diff
@@ -200,16 +208,75 @@ void LegoCarRaceActor::SwitchBoundary(LegoPathBoundary*& p_boundary, LegoUnknown
...
+		SwitchBoundary(m_boundary, m_destEdge, m_unk0xe4);
+		assert(m_boundary && m_destEdge);
+
+		// variable names verified by BETA10
+		Vector3* v1 = m_destEdge->CWVertex(*m_boundary);
+		Vector3* v2 = m_destEdge->CCWVertex(*m_boundary);
+		assert(v1 && v2);
+
+		Mx3DPointFloat point1;
+		point1[0] = (*v1)[0] + ((*v2)[0] - (*v1)[0]) * m_unk0xe4;
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/src/race/legoracespecial.cpp:247]: This should be a call to the `LERP3` operation/macro from `vec.h`

File: LEGO1/lego/legoomni/src/race/legoracespecial.cpp [hunk 2]
```diff
@@ -200,16 +208,75 @@ void LegoCarRaceActor::SwitchBoundary(LegoPathBoundary*& p_boundary, LegoUnknown
...
-// STUB: LEGO1 0x10080be0
+// FUNCTION: LEGO1 0x10080be0
+// FUNCTION: BETA10 0x100cdc54
 MxResult LegoCarRaceActor::VTable0x9c()
 {
-	// TODO
+	MxS32 iVar4 = VTable0x1c(m_boundary, m_destEdge);
+
+	if (iVar4) {
+		SwitchBoundary(m_boundary, m_destEdge, m_unk0xe4);
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/src/race/legoracespecial.cpp:238]: Before the call to `SwitchBoundary`, a copy of `m_boundary` is taken:

`LegoPathBoundary* b = m_boundary;`

File: LEGO1/lego/legoomni/src/race/legoracespecial.cpp [hunk 3]
```diff
@@ -200,16 +208,75 @@ void LegoCarRaceActor::SwitchBoundary(LegoPathBoundary*& p_boundary, LegoUnknown
...
+		point1[0] = (*v1)[0] + ((*v2)[0] - (*v1)[0]) * m_unk0xe4;
+		point1[1] = (*v1)[1] + ((*v2)[1] - (*v1)[1]) * m_unk0xe4;
+		point1[2] = (*v1)[2] + ((*v2)[2] - (*v1)[2]) * m_unk0xe4;
+
+		Mx3DPointFloat point2;
+		Mx3DPointFloat point3;
+		Mx3DPointFloat point4;
+		Mx3DPointFloat point5;
+
+		m_destEdge->FUN_1002ddc0(*m_boundary, point2);
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/src/race/legoracespecial.cpp:256]: This call is incorrect - it uses the copies of `m_boundary` and `m_destEdge` taken earlier:

`d->FUN_1002ddc0(*b, point2);`

The function will then match ~95%.

File: LEGO1/lego/legoomni/src/race/legoracespecial.cpp [hunk 4]
```diff
@@ -200,16 +208,75 @@ void LegoCarRaceActor::SwitchBoundary(LegoPathBoundary*& p_boundary, LegoUnknown
...
+	}
 }
 
-// STUB: LEGO1 0x10080be0
+// FUNCTION: LEGO1 0x10080be0
+// FUNCTION: BETA10 0x100cdc54
 MxResult LegoCarRaceActor::VTable0x9c()
 {
-	// TODO
+	MxS32 iVar4 = VTable0x1c(m_boundary, m_destEdge);
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/src/race/legoracespecial.cpp:235]: Before the call to `VTable0x1c`, a copy of `m_destEdge` is taken:

`LegoUnknown100db7f4* d = m_destEdge;`

File: LEGO1/lego/legoomni/src/race/legoracespecial.cpp [hunk 5]
```diff
@@ -200,16 +208,75 @@ void LegoCarRaceActor::SwitchBoundary(LegoPathBoundary*& p_boundary, LegoUnknown
...
+
+	if (iVar4) {
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/src/race/legoracespecial.cpp:237]: `iVar4` can go away, just place the call directly into the if-clause

File: LEGO1/lego/legoomni/src/race/legoracespecial.cpp [hunk 6]
```diff
@@ -200,16 +208,75 @@ void LegoCarRaceActor::SwitchBoundary(LegoPathBoundary*& p_boundary, LegoUnknown
...
+		point4.EqualsCross(&point2, m_boundary->GetUnknown0x14());
+		point5.EqualsCross(m_boundary->GetUnknown0x14(), &point3);
+
+		point4.Unitize();
+		point5.Unitize();
+
+		((Vector3*) &point4)->Mul(5.0f);
+		((Vector3*) &point5)->Mul(5.0f);
+
+		MxResult res = VTable0x80(Vector3(m_roi->GetWorldPosition()), point4, point1, point5);
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/src/race/legoracespecial.cpp:268]: You can remove the explicit construction of `Vector3` - it happens implicitly

File: LEGO1/lego/legoomni/src/race/legoracespecial.cpp (hunk 4 above)
Comment (jonschz) [reply LEGO1/lego/legoomni/src/race/legoracespecial.cpp]: Thanks for working this out! I didn't think that one of the function calls would use the variables while the other one would not.

File: LEGO1/lego/legoomni/src/race/legoracespecial.cpp [hunk 7]
```diff
@@ -7,15 +7,24 @@
 #include "misc.h"
 #include "mxmisc.h"
 #include "mxvariabletable.h"
+#include "vec.h"
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/src/race/legoracespecial.cpp:10]: should be <vec.h>

**Pull request merged**